File: src/services/gameservice.js
// 5 turns
// 250 character limit
// player 1: count turn
// every prompt will inc count turn
// we save 


import aiService from './aiService';
import prisma from './prismaClient';

class GameService {
    async createGameFromMatch(playerOneId, playerTwoId) {


        // Generate character and secret
        // Need to generate random template first
        const template = await this.generateTemplate();
        const templateId = template[0].id;
        const character = this.generateCharacter(template);
        const secret = this.generateSecret(template);

        return await prisma.game.create({
            data: {
                playerOneId,
                playerTwoId,
                templateId,
                generatedCharacter: character,
                generatedSecret: secret,
                status: 'IN_PROGRESS',
                phase: 'DEFENSE'
            }
        });
    }


    async submitTurn(gameId, playerId, message) {
        const game = await prisma.game.findUnique({
            where: { id: gameId },
            include: { template: true }
        });

        //Once this error is caught, send a message in the frontend to wait for their opponent
        const turnCount = await this.getTurnCount(gameId, playerId, game.phase);
        if (turnCount >= game.maxTurnsPerPhase) {
            throw new Error('Turn limit reached');
        }


        const aiResponse = await aiService.getResponse(
            game,
            playerId,
            message,
            game.phase
        );
        const turn = await prisma.gameTurn.create({
            data: {
                gameId,
                playerId,
                phase: game.phase,
                turnNumber: turnCount + 1,
                playerMessage: message,
                aiResponse
            }
        });

        await this.checkPhaseTransition(gameId);
        await this.checkGameEnd(gameId);
        return turn;
    }
    validateTurn(game, playerId, message) {
        if (game.status !== 'IN_PROGRESS') {
            throw new Error('Game not in progress');
        }

        //Need to do something about spaces or white space characters can probably trim
        if (message.length > game.maxCharsPerMessage) {
            throw new Error(`Message exceeds ${game.maxCharsPerMessage} characters`);
        }

        // Verify player is part of this game
        if (playerId !== game.playerOneId && playerId !== game.playerTwoId) {
            throw new Error('Player not in this game');
        }
    }
    async getTurnCount(gameId, playerId, phase) {
        return await prisma.gameTurn.count({
            where: {
                gameId,
                playerId,
                phase
            }
        });
    }
    async checkPhaseTransition(gameId) {
        const game = await prisma.game.findUnique({
            where: { id: gameId }
        });

        if (game.phase !== 'DEFENSE') return;


        //May have to change something here specifically for isdefendingboolean
        const p1Turns = await this.getTurnCount(gameId, game.playerOneId, 'DEFENSE');
        const p2Turns = await this.getTurnCount(gameId, game.playerTwoId, 'DEFENSE');

        if (p1Turns >= game.maxTurnsPerPhase && p2Turns >= game.maxTurnsPerPhase) {
            await this.transitionToAttack(gameId);
        }
    }
    async transitionToAttack(gameId) {
        const game = await prisma.game.findUnique({
            where: { id: gameId },
            include: { template: true }
        });

        //Should save costs? Only if summary is actually shorter than the entire list of prompt and response
        const p1Summary = await this.generateDefenseSummary(gameId, game.playerOneId);
        const p2Summary = await this.generateDefenseSummary(gameId, game.playerTwoId);
        await prisma.game.update({
            where: { id: gameId },
            data: {
                phase: 'ATTACK',
                playerOneDefenseSummary: p1Summary,
                playerTwoDefenseSummary: p2Summary
            }
        });
    }
    async generateDefenseSummary(gameId, playerId) {
        const turns = await prisma.gameTurn.findMany({
            where: {
                gameId,
                playerId,
                phase: 'DEFENSE'
            },
            orderBy: { turnNumber: 'asc' }
        });
        //Summarize defense, prompt and response list should be sent to wrapper "Can you summarize"
        return await aiService.summarizeDefense(turns);

    }
    async checkGameEnd(gameId) {
        const game = await prisma.game.findUnique({
            where: { id: gameId }
        });

        if (game.phase !== 'ATTACK') return;

        const p1AttackTurns = await this.getTurnCount(gameId, game.playerOneId, 'ATTACK');
        const p2AttackTurns = await this.getTurnCount(gameId, game.playerTwoId, 'ATTACK');

        // Both players finished their attack turns
        if (p1AttackTurns >= game.maxTurnsPerPhase && p2AttackTurns >= game.maxTurnsPerPhase) {
            await this.determineWinner(gameId);
        }
    }

    async determineWinner(gameId) {
        const game = await prisma.game.findUnique({
            where: { id: gameId }
        });
        //Get all the attack turns and then check if the users win
        //Must achieve some goal like revealing the secret or convincing ai to do something

        const p1AttackTurns = await prisma.gameTurn.findMany({
            where: {
                gameId,
                playerId: game.playerOneId,
                phase: 'ATTACK'
            },
            orderBy: { turnNumber: 'asc' }
        });

        const p2AttackTurns = await prisma.gameTurn.findMany({
            where: {
                gameId,
                playerId: game.playerTwoId,
                phase: 'ATTACK'
            },
            orderBy: { turnNumber: 'asc' }
        });
        // Check if Player 1 broke Player 2's AI
        const p1Succeeded = await aiService.checkSecretRevealed(
            game,
            game.playerTwoId, // P1 attacks P2's AI
            p1AttackTurns
        );

        // Check if Player 2 broke Player 1's AI
        const p2Succeeded = await aiService.checkSecretRevealed(
            game,
            game.playerOneId, // P2 attacks P1's AI
            p2AttackTurns
        );

        if (p1Succeeded && !p2Succeeded) {
            winnerId = game.playerOneId;
            endReason = 'SECRET_REVEALED';
        } else if (p2Succeeded && !p1Succeeded) {
            winnerId = game.playerTwoId;
            endReason = 'SECRET_REVEALED';
        } else if (p1Succeeded && p2Succeeded) {
            // Both broke it - it's a draw (or you could have tiebreaker logic)
            winnerId = undefined;
            endReason = 'DRAW';
        } else {
            // Neither broke it - both AIs held strong (draw)
            winnerId = undefined;
            endReason = 'DRAW';
        }
        await prisma.game.update({
            where: { id: gameId },
            data: {
                status: 'FINISHED',
                winnerId,
                endReason
            }
        });

    }
    //Can fully implement later, users can look at history of games wins and losses
    async getGame(gameId) {
        return await prisma.game.findUnique({
            where: { id: gameId },
            include: {
                playerOne: true,
                playerTwo: true,
                template: true,
                turns: {
                    orderBy: { createdAt: 'asc' }
                }
            }
        });
    }

    //Use of templates - Madlib style so we can keep some consistency
    generateCharacter(template) {
        // TODO: Use template to generate random character
        return "A suspicious guard";
    }


    generateSecret(template) {
        // TODO: Use template to generate random secret
        return "The password is blue42";
    }
    async generateTemplate() {
        const count = await prisma.scenarioTemplate.count();
        const random = Math.floor(Math.random() * count);
        const template = await prisma.template.findMany({
            skip: random,
            take: 1,
            select: { id: true }  // or templateId, depending on your schema
        });
        return template;
    }
}
export default GameService;

File: serverjava/src/main/java/com/backend/promptvprompt/services/GameService.java
package com.backend.promptvprompt.services;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Random;

import org.springframework.stereotype.Service;

import com.backend.promptvprompt.models.Game;
import com.backend.promptvprompt.models.GameEndReason;
import com.backend.promptvprompt.models.GamePhase;
import com.backend.promptvprompt.models.GameStatus;
import com.backend.promptvprompt.models.GameTurn;
import com.backend.promptvprompt.models.ScenarioTemplate;
import com.backend.promptvprompt.models.User;
import com.backend.promptvprompt.repos.GameRepo;
import com.backend.promptvprompt.repos.GameTurnRepo;
import com.backend.promptvprompt.repos.ScenarioTemplateRepo;
import com.backend.promptvprompt.repos.UserRepo;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class GameService {
    private final GameRepo gameRepo;
    private final GameTurnRepo gameTurnRepo;
    private final ScenarioTemplateRepo scenarioTemplateRepo;
    private final UserRepo userRepo;
    private final AiService aiService;
    private final Random random = new Random();

    @Transactional
    public Game createGameFromMatch(String playerOneId, String playerTwoId) {
        try {
            // Fetch actual User objects
            User playerOne = userRepo.findById(playerOneId)
                    .orElseThrow(() -> new RuntimeException("Player one not found"));
            User playerTwo = userRepo.findById(playerTwoId)
                    .orElseThrow(() -> new RuntimeException("Player two not found"));

            ScenarioTemplate template = generateTemplate();
            String character = generateCharacter(template);
            String secret = generateSecret(template);

            Game game = Game.builder()
                    .playerOne(playerOne)
                    .playerTwo(playerTwo)
                    .template(template)
                    .generatedCharacter(character)
                    .generatedSecret(secret)
                    .status(GameStatus.DEFENSE_PHASE)
                    .phase(GamePhase.DEFENSE)
                    .build();

            return gameRepo.save(game);
        } catch (Exception e) {
            throw new RuntimeException("Error creating game from match: " + e.getMessage(), e);
        }
    }

    public GamePhase getPhase(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));
        return game.getPhase();
    }

    @Transactional
    public GameTurn submitTurn(String gameId, String playerId, String message) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        validateTurn(game, playerId, message);

        int turnCount = getTurnCount(gameId, playerId, game.getPhase());

        if (turnCount >= game.getMaxTurnsPerPhase()) {
            throw new IllegalStateException("Turn limit reached");
        }

        String aiResponse = aiService.getResponse(game, playerId, message, game.getPhase());
        User player = userRepo.findById(playerId)
                .orElseThrow(() -> new RuntimeException("Player not found"));
        GameTurn turn = GameTurn.builder()
                .game(game)
                .player(player)
                .phase(game.getPhase())
                .turnNumber(turnCount + 1)
                .playerMessage(message)
                .aiResponse(aiResponse)
                .build();

        GameTurn savedTurn = gameTurnRepo.save(turn);

        checkPhaseTransition(gameId);
        checkGameEnd(gameId);

        return savedTurn;
    }

    void validateTurn(Game game, String playerId, String message) {
        if (game.getStatus() != GameStatus.ATTACK_PHASE &&
                game.getStatus() != GameStatus.DEFENSE_PHASE) {
            throw new IllegalStateException("Game not in progress");
        }

        String trimmedMessage = message.trim();
        if (trimmedMessage.length() > game.getMaxCharsPerMessage()) {
            throw new IllegalArgumentException(
                    String.format("Message exceeds %d characters", game.getMaxCharsPerMessage()));
        }

        if (!playerId.equals(game.getPlayerOne().getId()) &&
                !playerId.equals(game.getPlayerTwo().getId())) {
            throw new IllegalArgumentException("Player not in this game");
        }
    }

    public int getTurnCount(String gameId, String playerId, GamePhase phase) {
        return gameTurnRepo.countByGameIdAndPlayerIdAndPhase(gameId, playerId, phase);
    }

    public List<GameTurn> getTurns(String gameId, String playerId, GamePhase phase) {
        return gameTurnRepo.findByGameIdAndPlayerIdAndPhaseOrderByTurnNumberAsc(
                gameId, playerId, phase);
    }

    @Transactional
    public void checkPhaseTransition(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        if (game.getPhase() != GamePhase.DEFENSE) {
            return;
        }

        int p1Turns = getTurnCount(gameId, game.getPlayerOne().getId(), GamePhase.DEFENSE);
        int p2Turns = getTurnCount(gameId, game.getPlayerTwo().getId(), GamePhase.DEFENSE);

        if (p1Turns >= game.getMaxTurnsPerPhase() &&
                p2Turns >= game.getMaxTurnsPerPhase()) {
            transitionToAttack(gameId);
        }
    }

    @Transactional
    public void transitionToAttack(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        String p1Summary = generateDefenseSummary(gameId, game.getPlayerOne().getId());
        String p2Summary = generateDefenseSummary(gameId, game.getPlayerTwo().getId());

        game.setStatus(GameStatus.ATTACK_PHASE);
        game.setPhase(GamePhase.ATTACK);
        game.setPlayerOneDefenseSummary(p1Summary);
        game.setPlayerTwoDefenseSummary(p2Summary);
        game.setIsTransitioning(true);
        game.setTransitionEndsAt(LocalDateTime.now().plusSeconds(5));

        gameRepo.save(game);
    }

    @Transactional
    public void endTransition(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        game.setIsTransitioning(false);
        game.setTransitionEndsAt(null);

        gameRepo.save(game);
    }

    public String generateDefenseSummary(String gameId, String playerId) {
        List<GameTurn> turns = gameTurnRepo
                .findByGameIdAndPlayerIdAndPhaseOrderByTurnNumberAsc(
                        gameId, playerId, GamePhase.DEFENSE);

        return aiService.summarizeDefense(turns);
    }

    @Transactional
    public boolean checkGameEnd(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        if (game.getPhase() != GamePhase.ATTACK) {
            return false;
        }

        int p1AttackTurns = getTurnCount(gameId, game.getPlayerOne().getId(), GamePhase.ATTACK);
        int p2AttackTurns = getTurnCount(gameId, game.getPlayerTwo().getId(), GamePhase.ATTACK);

        if (p1AttackTurns >= game.getMaxTurnsPerPhase() &&
                p2AttackTurns >= game.getMaxTurnsPerPhase()) {
            determineWinner(gameId);
            return true;
        }
        return false;
    }

    @Transactional
    public void determineWinner(String gameId) {
        Game game = gameRepo.findById(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));

        User winner = null;
        GameEndReason endReason = GameEndReason.DRAW;

        List<GameTurn> p1AttackTurns = gameTurnRepo
                .findByGameIdAndPlayerIdAndPhaseOrderByTurnNumberAsc(
                        gameId, game.getPlayerOne().getId(), GamePhase.ATTACK);

        List<GameTurn> p2AttackTurns = gameTurnRepo
                .findByGameIdAndPlayerIdAndPhaseOrderByTurnNumberAsc(
                        gameId, game.getPlayerTwo().getId(), GamePhase.ATTACK);

        boolean p1Succeeded = aiService.checkSecretRevealed(
                game, p1AttackTurns);

        boolean p2Succeeded = aiService.checkSecretRevealed(
                game, p2AttackTurns);

        if (p1Succeeded && !p2Succeeded) {
            winner = game.getPlayerOne();
            endReason = GameEndReason.FULL_CONVICTION;
        } else if (p2Succeeded && !p1Succeeded) {
            winner = game.getPlayerTwo();
            endReason = GameEndReason.FULL_CONVICTION;
        } else if (p1Succeeded && p2Succeeded) {
            winner = null;
            endReason = GameEndReason.DRAW;
        } else {
            winner = null;
            endReason = GameEndReason.DRAW;
        }

        game.setStatus(GameStatus.COMPLETED);
        game.setWinner(winner);
        game.setEndReason(endReason);

        gameRepo.save(game);
    }

    public Game getGame(String gameId) {
        return gameRepo.findByIdWithTemplateAndTurns(gameId)
                .orElseThrow(() -> new RuntimeException("Game not found"));
    }

    String generateCharacter(ScenarioTemplate template) {
        @SuppressWarnings("unchecked")
        List<String> roles = (List<String>) template.getVariables().get("role");
        int randomIndex = random.nextInt(roles.size());
        return roles.get(randomIndex);
    }

    String generateSecret(ScenarioTemplate template) {
        @SuppressWarnings("unchecked")
        List<String> passwords = (List<String>) template.getVariables().get("password");
        int randomIndex = random.nextInt(passwords.size());
        return passwords.get(randomIndex);
    }

    ScenarioTemplate generateTemplate() {
        long count = scenarioTemplateRepo.count();
        if (count == 0) {
            throw new RuntimeException("No templates available");
        }

        int randomIndex = random.nextInt((int) count);
        return scenarioTemplateRepo.findRandomTemplate(randomIndex)
                .orElseThrow(() -> new RuntimeException("No template found"));
    }
}